Allow More Child Processes

NoMoreFood opened the pull request:
- Added wait_for_multiple_objects_enhanced() function that can logically handle wait-any on many children (system resources permitting).
- Added unit tests for wait_for_multiple_objects_enhanced().
- Incorporated use of wait_for_multiple_objects_enhanced() into main signal handler.
- Upped max child processes to 500.

NoMoreFood committed:
Allow More Child Processes

- Added wait_for_multiple_objects_enhanced() function that can logically handle wait-any on many children (system resources permitting).
- Added unit tests for wait_for_multiple_objects_enhanced().
- Incorporated use of wait_for_multiple_objects_enhanced() into main signal handler.
- Upped max child processes to 500.

NoMoreFood committed:
Allow More Child Processes - Visual Studio Project

- Added signal_wait.c and signal_tests.c to the Visual Studio project files.

manojampalam commented on contrib/win32/win32compat/signal_wait.c:
```diff
@@ -0,0 +1,194 @@
+	wait_for_multiple_objects_struct * waitstruct = (wait_for_multiple_objects_struct *) lpParam;
+
+	/* wait for bin to complete -- this is alertable for our interrupt cleanup routine */
+	waitstruct->return_value = WaitForMultipleObjectsEx(waitstruct->num_handles,
```
assuming the initialized value of waitstruct->return_value is 0, this logic cannot distinguish if the WaitForMultipleObjectsEx successfully completes with WAIT_OBJECT_0 (its value is 0 too).

manojampalam replied:
Never mind.. I see that return_value is initialized to WAIT_FAILED - 1

manojampalam commented on contrib/win32/win32compat/signal_wait.c:
```diff
@@ -0,0 +1,194 @@
+	const DWORD bins_total = (nCount - 1) / bin_size + 1;
+
+	/* setup synchronization variables */
+	HANDLE wait_event = CreateEvent(NULL, TRUE, FALSE, NULL);
```
handle CreateEvent failure

manojampalam commented on contrib/win32/win32compat/signal_wait.c:
```diff
@@ -0,0 +1,194 @@
+
+	/* allocate an area to communicate with our threads */
+	wait_for_multiple_objects_struct * wait_bins = (wait_for_multiple_objects_struct *)
+		calloc(bins_total, sizeof(wait_for_multiple_objects_struct));
```
handle calloc failure

manojampalam commented on contrib/win32/win32compat/signal_wait.c:
```diff
@@ -0,0 +1,194 @@
+	/* wait for at least one thread to return */
+	DWORD return_value = WAIT_FAILED_ENHANCED;
+	DWORD wait_ret = WaitForSingleObjectEx(wait_event, dwMilliseconds, bAlertable);
+
```
You should issue APCs on worker threads right here and wait for all of them to terminate. This needs to be done before you read "bins" state below (to synchronize writes and reads on this state).

NoMoreFood replied:
The only potential conflict value would be return_value which would be an aligned 32-bit value.  Those read/writes will always be atomic and at least one of them will be set before SetEvent() is called.   Am I overlooking another potential concern?

manojampalam replied:
No. There isn't any potential concern yet. From a logical stand point, once the main thread woke up, there is no point keeping the worker threads active. If we abort and clean them up right away, the rest of the trailing code need not make any assumptions of what else might be running in parallel. This will help us keep the logic simple, easy to maintain and debug (in case of any issues).

manojampalam reviewed (changes requested):
Added comments. 
Overall, your code looks clean, but please refer to the OpenBSD coding guidelines once and ensure the indentation rules are followed. We eventually want to integrate this fork into upstream repo that thoroughly follows OpenBSD coding standards.

NoMoreFood commented:
@manojampalam Thank you for the review.  Is there a particular place or two you can highlight where I'm off-base with the coding style?

manojampalam commented:
OpenBSD guidelines are at htttp://man.openbsd.org/cgi-bin/man.cgi/OpenBSD-current/man9/style.9?query=style%26sec=9.

NoMoreFood committed:
Error Handling; Updated Stytle

- Added additional code for error handling during allocations.
- Updated coding to be consistent with style from upstream.
- Updated unit tests for wait-any version of wait_for_multiple_objects_enhanced().

NoMoreFood commented:
I just pushed another commit to address most of your comments.  I could not cleanly move the thread cleanup code since the possibility existed that a CreateThread() could fail in the beginning in which case there would not be an elegant way to do thread cleanup without unnecessary code duplication.  However, I updated comments to hopefully give the reader a better understanding of the execution.  I also had mistakenly submitted an earlier version of the my unit test cases in my previous commit that is why there are so many changes in that file.

manojampalam commented on contrib/win32/win32compat/signal_wait.c:
```diff
@@ -0,0 +1,220 @@
+	if (wait_bins == NULL) {
+		goto cleanup;
+	}
+
```
ZeroMemory(wait_bins), otherwise, if any thread creation fails below, you will be dealing with uninitialized memory in cleanup

manojampalam commented on contrib/win32/win32compat/signal_wait.c:
```diff
@@ -0,0 +1,220 @@
+	}
+
+	/* number of separate bins / threads required to monitor execution */
+	const DWORD bin_size = MAXIMUM_WAIT_OBJECTS;
```
Do all declarations at start of code block and initialize them.

manojampalam committed:
Minor changes to wait_for_multiple_objects_enhanced

Moved variable declarations to top of code block
ZeroMemory'd allocated memory

manojampalam commented:
I've added a commit with minor changes in. I'll merge your changes once the validation is done.
